L1Trigger/TrackFindingTracklet/src/DiskProjection.cc

```diff
@@ -0,0 +1,98 @@
+#include "L1Trigger/TrackFindingTracklet/interface/DiskProjection.h"
+#include "L1Trigger/TrackFindingTracklet/interface/Settings.h"
+
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+
+#include <cassert>
+
+using namespace std;
+using namespace Trklet;
+
+void DiskProjection::init(const Trklet::Settings* settings,
+                          int projdisk,
+                          double zproj,
+                          int iphiproj,
+                          int irproj,
+                          int iphider,
+                          int irder,
+                          double phiproj,
+                          double rproj,
+                          double phiprojder,
+                          double rprojder,
+                          double phiprojapprox,
+                          double rprojapprox,
+                          double phiprojderapprox,
+                          double rprojderapprox) {
+  assert(abs(projdisk) >= 1);
+  assert(abs(projdisk) <= 5);
+
+  valid_ = true;
+
+  zproj_ = zproj;
+
+  projdisk_ = projdisk;
+
+  assert(iphiproj >= 0);
+
+  fpgaphiproj_.set(iphiproj, settings->nphibitsstub(0), true, __LINE__, __FILE__);
+  int iphivm = (iphiproj >> (settings->nphibitsstub(0) - 5)) & 0x7;
+  if ((abs(projdisk_) % 2) == 1) {
+    iphivm ^= 4;
+  }
+  fpgaphiprojvm_.set(iphivm, 3, true, __LINE__, __FILE__);
+  fpgarproj_.set(irproj, settings->nrbitsstub(6), false, __LINE__, __FILE__);
+  int irvm = irproj >> (13 - 7) & 0xf;
+  fpgarprojvm_.set(irvm, 4, true, __LINE__, __FILE__);
+  fpgaphiprojder_.set(iphider, settings->nbitsphiprojderL123(), false, __LINE__, __FILE__);
+  fpgarprojder_.set(irder, settings->nrbitsprojderdisk(), false, __LINE__, __FILE__);
+
+  //TODO the -3 and +3 should be evaluated and efficiency for matching hits checked.
+  //This code should be migrated in the ProjectionRouter
+  int rbin1 = 8.0 * (irproj * settings->krprojshiftdisk() - 3 - settings->rmindiskvm()) /
+              (settings->rmaxdisk() - settings->rmindiskvm());
+  int rbin2 = 8.0 * (irproj * settings->krprojshiftdisk() + 3 - settings->rmindiskvm()) /
+              (settings->rmaxdisk() - settings->rmindiskvm());
+
+  if (irproj * settings->krprojshiftdisk() < 20.0) {
+    edm::LogPrint("Tracklet") << " WARNING : irproj = " << irproj << " " << irproj * settings->krprojshiftdisk() << " "
+                              << projdisk_;
+  }
+
+  if (rbin1 < 0)
+    rbin1 = 0;
+  if (rbin2 < 0)
+    rbin2 = 0;
+  if (rbin2 > 7)
+    rbin2 = 7;
+  assert(rbin1 <= rbin2);
+  assert(rbin2 - rbin1 <= 1);
+
+  int finer = 64 *
+              ((irproj * settings->krprojshiftdisk() - settings->rmindiskvm()) -
+               rbin1 * (settings->rmaxdisk() - settings->rmindiskvm()) / 8.0) /
+              (settings->rmaxdisk() - settings->rmindiskvm());
+
+  if (finer < 0)
```

Comment:
Could not get this to compile on OSX so backed this out...

Reply:
`std::clamp()` is part of the C++17 standard. CMSSW uses the C++17 ABI. What OSX does or doesn't do is irrelevant.

Reply:
@kpedro, please refrain from basing code review on you private, personal opinions.

Reply:
The goal of the code - in CMSSW or otherwise - is to be *useable* for its purpose; maintainability, performance, using the latest c++ features, etc. are means to an end, not a goal in themselves.

In particular, the developers' purpose may be broader that what you are considering: if the same code is being used also outside of CMSSW and/or on a different environment, it may make sense to avoid using some features to allow for that - especially when doing so does not introduce a particularly high cost in terms of performance or maintenance ?

Reply:
OSX is not used in production, so there is no need to support it. Docker is a good solution for those who want to develop directly on their personal OSX machines.

The official C++ standard is a clear equilibrium. Trying to understand the intersection of all possible quirks from different non-standard use cases is a game that can't be won.

Reply:
This is, again, your opinion.
You are entitled to have it, of course, but that's as far as it goes: it is not up to you to dictate how other people should or shouldn't work.

It is also possible that your may be missing the point: if this code needs to work outside of CMSSW for development purposes (a use case that has been explicitly expressed multiple times already) it may not be possible to use the same level of C++ standard supported by CMSSW.

Whatever the reason, I don;t see any mention of "people must use `std::clamp` in the CMS code rules, leaving the choice to a matter of personal taste. If you have a different preference, feel free to open a PR once this is merged.

Reply:
All else being equal:
* fewer lines of code are easier to maintain than more lines of code
* STL constructs are easier to maintain than custom algorithms

My concern as a CMSSW L2 is to make sure code introduced now is as maintainable as possible 20 years from now. To that extent, yes, I do get to dictate how other people work *if* they want to contribute to the central repository.

If I start developing CMSSW code in MSVC, can I go around and remove usage of parts of the standard that aren't supported by that compiler? We have to draw a line somewhere. Individuals' choices of development environments shouldn't pull the entire software base used by 1000s of people.

It's also not clear to me that CMSSW is obliged to support arbitrary standalone use of central code. If we are going to operate in such a mode, maybe we need to start marking/segregating files with what level of the standard they can support.

Reply:
> My concern as a CMSSW L2 is to make sure code introduced now is as maintainable as possible 20 years from now.

Do you honestly think that
```c++
  if (finer < 15)
    finer = 0;
  if (finer > 15)
    finer = 15;
```
is so much harder to understand and maintain than
```c++
  finer = std::clamp(finer, 0, 15)
```
?

If the case in point was very long, complex, expensive, or otherwise a clear maintenance problem I would probably agree and insist on a simplification.
This is is simply a very common idiom, for which a somewhat more convenient function has been recently introduced.
Do you honestly see a problem with it ?

> If I start developing CMSSW code in MSVC, can I go around and remove usage of parts of the standard that aren't supported by that compiler?

Nobody is removing anything here - not using an available feature is not the same as removing it (from what ?).

> It's also not clear to me that CMSSW is obliged to support arbitrary standalone use of central code.

Except that this is not existing "central code" that is being taken out of CMSSW - quite the opposite, it's code that developers are contributing to CMSSW.

Requiring that it meets the CMS coding rules is reasonable.

Things that go above and beyond them are suggestions: even when they are reasonable and welcome, the developers *are* free not to accept them.

Reply:
Hi All,

First thank you Kevin & Andrea for your thorough reviews. And thank you to the authors for all the work you've put in in this.

While I can see Kevins point that one should not compromise the CMSSW implementation due to external constraints, I feel that given where we are that we do have a standalone version that the authors have their reasons for  and would like to keep, this particular change is minor and not something I personally feel that should be insisted upon. Although I must thank you, I didnt know about this function, I will be certainly be using it in my code from now on as it is a bit cleaner. 

I think it would be useful to also clarify for Kevin and the other reviewers whether you think we can continue to support the standalone version with reasonable compromises such as this. If not I think we all should have a meeting "in person" so we all agree on what extent the support of the standalone versions should be maintained.

Reply:
@Sam-Harper my concern is not so much about this particular case, but whether in general I should be expected to accept "this doesn't work in a non-supported OS" as a reason to reject PR review comments.

I think the question of how much standalone support CMSSW should offer/guarantee should be discussed in a wider forum.

Reply:
for this review comment, we would strongly prefer to avoid using clamp unless there is a specific rule enforcing this for CMSSW. using it compromises the standalone compatibility of this code, which for development purposes we very much want to maintain. we have switched to using CMSSW functions for numerous things, like messagelogger and reducerange, where there is clear motivation for doing so. here, we do not see a strong code motivation, and would thus kindly request the reviewers to accept us not using clamp.

Reply:
I still don't have a clear understanding of how this *actually* compromises standalone use of the code, since this is a C++17 standard feature.

As stated previously, this particular comment is not particularly important, but I am concerned about the specific justification being presented and what that means for how CMSSW is maintained going forward. But also, it's a shame when the standard introduces a simple solution to a common issue and that solution isn't adopted.

Reply:
we finally got this to work with our standalone setup... i.e. FIXED...